activesupport/lib/active_support/message_verifier.rb

```diff
@@ -64,54 +66,156 @@ def valid_message?(signed_message)
+    # of the previous rails version.
     #
     #   verifier = ActiveSupport::MessageVerifier.new 's3Krit'
-    #   signed_message = verifier.generate 'a private message'
+    #   signed_message = verifier.generate('a private message', for: 'test', expires_in: 1.day)
+    #
+    #   #within 1 day...
+    #   verifier.verify(signed_message, for: 'test') # => "a private message"
     #
-    #   verifier.verify(signed_message) # => 'a private message'
+    #   legacy_signed_message = "BAhJIhZhIHByaXZhdGUgbWVzc2FnZQY6BkVU--a2a1232deaf49ba164ed0d0b43b3f17cf8026c46"
+    #   verifier.verify(legacy_signed_message) # => "a private message"
     #
     # Raises +InvalidSignature+ if the message was not signed with the same
     # secret or was not Base64-encoded.
     #
     #   other_verifier = ActiveSupport::MessageVerifier.new 'd1ff3r3nt-s3Krit'
     #   other_verifier.verify(signed_message) # => ActiveSupport::MessageVerifier::InvalidSignature
-    def verify(signed_message)
-      verified(signed_message) || raise(InvalidSignature)
+    def verify(signed_message, options = {})
+      verified(signed_message, options) || raise(InvalidSignature)
     end
 
-    # Generates a signed message for the provided value.
+    # Generates a signed message for the given +value+ and +options+ hash.
     #
     # The message is signed with the +MessageVerifier+'s secret. Without knowing
-    # the secret, the original value cannot be extracted from the message.
+    # the secret, the original value cannot be extracted from the message. A
+    # message is signed in the form of JSON Web Token(JWT). The final message is
+    # concatenation of base64-encoded JWS header, base64-encoded JWT claims set
+    # and  base64-encoded signature seperated by a period('.'). This format is
+    # known as JSON Web Signature(JWS) structure:
+    # <base64-encoded header>.<base64-encoded claims>.<base64-encoded signature>
+    #
+    # JSON Web Token (JWT)  A string representing a set of claims as a JSON
+    # object that is encoded in a JWS or JWE, enabling the claims to be
+    # signed and/or encrypted.
+    #
+    # JWT Header  A JSON object that describes the cryptographic operations
+    # applied to the JWT.  When the JWT is signed, the JWT Header is a
+    # JWS Header. A JWS header signed using 'SHA1' algorithm:
+    #
+    #   { 'typ' => 'JWT', 'alg' => 'SHA1' }
+    #
+    # JWT Claims Set  A JSON object that contains the Claims conveyed by
+    # the JWT. For our purpose, we limit our claims set to consist of
+    # +pld+(payload), +exp+(expiry), and +for+(purpose/scope). This becomes
+    # the payload of JWS. For example:
+    #
+    #   { 'pld' => 'a private message', 'for' => 'test', 'exp' => '2015-06-29T07:16:12.579Z' }
+    #
+    # JWS Signature  Signature over the JWS Header and the JWS Payload(Claims set).
+    # The signing input is given by ASCII(BASE64URL(UTF8(JWS Protected Header)) || '.' ||
+    # BASE64URL(JWS Payload)).
```

Comment:
I have not done strictly UTF8 of the JWS protected header as written above. As far as I know, serialized(e.g Marshal.dump) data is also a  UTF8 string. Here http://tools.ietf.org/html/rfc7515#section-3.1